### DISCUSSION TITLE
Parallel yaw optimization under uncertainty

### DISCUSSION THREAD

**Discussion by MiguelMarante** [NatLabRockies/floris#1002] [General]
Hello everyone,

I am trying to run optimization in parallel under uncertainty. However, I have noticed that the output differs significantly when running in parallel. It appears that the optimization is performed independently for each sampled wind direction.

Below is a simple example using a wind direction standard deviation of 3. When running the optimization in series, the results show the optimal yaw set-points for the mean wind direction, as expected. However, when running in parallel, these set-points are provided for each of the sampled wind directions ([264., 267., 270., 273., 276.]).

I wonder if this is a problem in my implementation or in the coupling done between the Uncertain and Parallel classes in Floris. Is there any workaround for this problem?

Best regards,
Miguel Marante

CODE:
```python
import numpy as np
from floris import WindRose, UncertainFlorisModel, ParallelFlorisModel
from floris.optimization.yaw_optimization.yaw_optimizer_sr import YawOptimizationSR

# Input model config
input_file = 'floris/examples/inputs/gch.yaml'

# Input climate data
ti = 0.06
ws = np.array([6., 8., 10.])
wd = np.array([270.])
wind_rose = WindRose(
    wind_speeds=ws,
    wind_directions=wd,
    ti_table=ti
    )

# Create floris object
ufmodel = UncertainFlorisModel(input_file, wd_std=3.0)
ufmodel.set(wind_data=wind_rose)

# Run optimization
yaw_opt = YawOptimizationSR(ufmodel)
df_opt_series = yaw_opt.optimize()
print(df_opt_series)


pfmodel = ParallelFlorisModel(ufmodel, max_workers=3, n_wind_condition_splits=3)
df_opt_parallel = pfmodel.optimize_yaw_angles()
print(df_opt_parallel)
```

![image](https://github.com/user-attachments/assets/4b240f44-8fdc-4dff-9aa9-5e0a78ccad37)

**Comment by misi9170** [comment]
Hi @MiguelMarante ,

Thanks for your question! We have actually very recently introduced a new parallel option, the `ParFlorisModel`, in [#982 ](https://github.com/NREL/floris/pull/982). This is currently only available on FLORIS' develop branch, but it will be available in the next week or so on the main branch also (we are planning a new minor version release, 4.2, which we are just putting the finishing touches on---see https://github.com/NREL/floris/pull/994). 

To avoid breaking changes, we will not remove the existing `ParallelFlorisModel`---however, we have found that there is not good compatibility between `UncertainFlorisModel` and `ParallelFlorisModel` (which we have improved now with the new `ParFlorisModel`). To avoid confusion, in v4.2 you won't be allowed to pass the old-style `ParallelFlorisModel` into `UncertainFlorisModel` any longer, because it doesn't work properly anyway.

Note that the new `ParFlorisModel` does not have it's own `optimize_yaw_angles()` method. Rather, an instantiated `ParFlorisModel` object should be passed directly to `YawOptimizerSR`, in exactly the same way as an `UncertainFlorisModel` would be. In your case, the workflow would be:
1. Instantiate a `ParFlorisModel` (e.g., named `pfmodel`)
2. Instantiate an `UncertainFlorisModel` (e.g., named `ufmodel`), passing in `pfmodel` to the `configuration` argument
3. Instantiate a yaw optimizer object `yaw_opt = YawOptimizationSR(ufmodel)` and proceed to optimize.

Hopefully that makes sense---and sorry to switch this up on you---we hope that you'll find that the new `ParFlorisModel` is more directly interchangeable with `FlorisModel` than the previous `ParallelFlorisModel`!

Misha

**Reply by paulf81** [reply]
Just to add on, using the newer syntax this is working properly I think:

```python
import numpy as np
import pandas as pd
from floris import WindRose, UncertainFlorisModel, ParFlorisModel
from floris.optimization.yaw_optimization.yaw_optimizer_sr import YawOptimizationSR

# Included for parallelization
if __name__ == "__main__":

    # Input model config
    input_file = 'inputs/gch.yaml'

    # Input climate data
    ti = 0.06
    ws = np.array([6., 8., 10.])
    wd = np.array([270.])
    wind_rose = WindRose(
        wind_speeds=ws,
        wind_directions=wd,
        ti_table=ti
        )

    # Create uncertain floris object
    ufmodel = UncertainFlorisModel(input_file, wd_std=3.0)
    ufmodel.set(wind_data=wind_rose)

    # Create parallelelized uncertain floris object
    pfmodel = ParFlorisModel(input_file, max_workers=3, n_wind_condition_splits=3)
    upfmodel = UncertainFlorisModel(pfmodel, wd_std=3.0)

    # Run optimization
    yaw_opt = YawOptimizationSR(ufmodel)
    df_opt_series = yaw_opt.optimize()

    # Run optimization with parallelization
    yaw_opt_par = YawOptimizationSR(upfmodel)
    df_opt_series_par = yaw_opt.optimize()

    # Print results
    print(df_opt_series)
    print('Parallelized results:')
    print(df_opt_series_par)

```

![image](https://github.com/user-attachments/assets/66e336eb-0742-4c17-a6be-4b8bdd7ed59e)

**Reply by MiguelMarante** [reply]
Hello. Thank you, I am glad to hear that you already fixed this. Perfect timing :)
